README.md

```diff
@@ -1,55 +1,385 @@
+segmentation fault like it could happen in other languages.
+
+# Usage
+
+This part provides notes about how to use the provided data structures or more
+exactly, what to not do with it. Two main points are discussed:
+
+- some data structures have restrictions on one domain can do
+- the currently provided data structures are non-composable
+
+## Data structures with domain roles
+
+There are several provided data structures that are meant to be used with a
+certain domain configuration. These restrictions enable to do optimizations but
+that breaks safety properties if not respected. Obviously, these restrictions
+are not only written in documentation but also on the name of the data structure
+itself. Single consumer queue means only one domain should pop.
+
+Let's take the example of `Single_prod_single_cons_queue`. As suggested by the
+name, it should be used with only one domain performing `push` (a producer) and
+one domain performing `pop` (a consumer) at the same time. Having two or more
+domains try to perform one of these operations at the same time will break
+safety properties of the `queue` and more likely result in unexpected
+behaviours.
+
+Let's say we give a bad alias to this queue and misuse it.
+
+```ocaml
+module Queue = Saturn.Single_prod_single_cons_queue
+```
+
+Each domain is going to try to push 10 times in parallel.
+
+```ocaml
+let work id barrier q =
+  Barrier.await barrier;
+  for i = 0 to 9 do
+    Queue.push q id
+  done
+```
+
+Our `test` function returns the queue after two domains try to simustaneously
+push.
+
+```ocaml
+let test () =
+  let q = Queue.create ~size_exponent:5 in
+  let barrier = Barrier.create 2 in
+  let d1 = Domain.spawn (fun () -> work 1 barrier q) in
+  let d2 = Domain.spawn (fun () -> work 2 barrier q) in
+  Domain.join d1;
+  Domain.join d2;
+  q
+```
+
+We can then inspect the queue content by popping it into a list.
+
+```ocaml
+let get_content q =
+  let rec loop acc =
+    match Queue.pop q with
+    | None -> acc
+    | Some a -> loop (a :: acc)
+  in
+  loop [] |> List.rev
+```
+
+Let's run it :
+
+```ocaml
+test () |> get_content;;
+- : int list = [2; 1; 1; 1; 1; 1; 1; 1; 1; 1; 2]
+```
+
+This run results in a queue with only 11 elements. 9 elements are lost to the
+misuse of the single consumer single producer queue.
+
+## About composability
+
+TODO
+
+# Properties and testing
+
+## Liveness properties
+
+One of the many difficulties of implementating parallelism-safe data structures
+is that in addition to providing the same safety properties as sequental one,
+they may also have to observe some _liveness_ properties.
+
 ## Tests
 
 Several kind of tests are provided for each data structure:
 
-- unitary tests and `qcheck` tests: check semantics and expected behaviors with
-  one and more domains;
-- `STM` tests: check _linearizability_ for two domains (see
+- _safety_ : unitary tests and `qcheck` tests: check semantics and expected
```

Comment:
```suggestion
- _safety_ : unitary tests and `qcheck` tests check semantics and expected
```
I'm not sure you need a colon here since "check" works as a verb.